File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    int result = 0;

    long dfs(int curr, int parent, unordered_map<int, vector<int>>& adj, vector<int>& values, int k){
        long total = values[curr];

        for (int node: adj[curr])
            if (node != parent)
                total += dfs(node, curr, adj, values, k);

        if (total % k == 0){
            result += 1;
        }

        return total;
    }

    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        unordered_map<int, vector<int>> adj;
        for (vector<int> edge: edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        dfs(0, -1, adj, values, k);
        return result;
    }
};

File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        adjMatrix = defaultdict(list)
        for i, j in edges:
            adjMatrix[i].append(j)
            adjMatrix[j].append(i)
        print(adjMatrix)
        result = 0

        def dfs(curr, parent):
            nonlocal result
            total = values[curr]

            for node in adjMatrix[curr]:
                if node != parent:
                    total += dfs(node, curr)

            if total % k == 0:
                result += 1

            return total
        
        dfs(0, -1)
        return result